ext/opentelemetry-ext-botocore/src/opentelemetry/ext/botocore/__init__.py

```diff
@@ -0,0 +1,208 @@
+    )
+    ec2 = self.session.create_client("ec2", region_name="us-west-2")
+    ec2.describe_instances()
+
+API
+---
+"""
+
+import logging
+
+from botocore.client import BaseClient
+from wrapt import ObjectProxy, wrap_function_wrapper
+
+from opentelemetry.auto_instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.ext.botocore.version import __version__
+from opentelemetry.trace import SpanKind, get_tracer
+
+logger = logging.getLogger(__name__)
+
+ARGS_NAME = ("action", "params", "path", "verb")
+TRACED_ARGS = ["params", "path", "verb"]
+
+
+class BotocoreInstrumentor(BaseInstrumentor):
+    """A instrumentor for Botocore
+
+    See `BaseInstrumentor`
+    """
+
+    def _instrument(self, **kwargs):
+
+        # FIXME should the tracer provider be accessed via Configuration
+        # instead?
+        # pylint: disable=attribute-defined-outside-init
+        self._tracer = get_tracer(
+            __name__, __version__, kwargs.get("tracer_provider")
+        )
+
+        wrap_function_wrapper(
+            "botocore.client",
+            "BaseClient._make_api_call",
+            self._patched_api_call,
+        )
+
+    def _uninstrument(self, **kwargs):
+        unwrap(BaseClient, "_make_api_call")
+
+    def _patched_api_call(self, original_func, instance, args, kwargs):
+
+        endpoint_name = deep_getattr(instance, "_endpoint._endpoint_prefix")
+
+        with self._tracer.start_as_current_span(
+            "{}.command".format(endpoint_name), kind=SpanKind.CONSUMER,
+        ) as span:
+
+            operation = None
+            if args:
+                operation = args[0]
+                span.resource = "%s.%s" % (endpoint_name, operation.lower())
+
+            else:
+                span.resource = endpoint_name
+
+            add_span_arg_tags(
+                span, endpoint_name, args, ARGS_NAME, TRACED_ARGS
+            )
+
+            region_name = deep_getattr(instance, "meta.region_name")
+
+            meta = {
+                "aws.agent": "botocore",
+                "aws.operation": operation,
+                "aws.region": region_name,
+            }
+            for key, value in meta.items():
+                span.set_attribute(key, value)
+
+            result = original_func(*args, **kwargs)
+
+            span.set_attribute(
+                "http.status_code",
+                result["ResponseMetadata"]["HTTPStatusCode"],
+            )
+            span.set_attribute(
+                "retry_attempts", result["ResponseMetadata"]["RetryAttempts"],
+            )
+
+            return result
+
+
+def unwrap(obj, attr):
+    function = getattr(obj, attr, None)
+    if (
+        function
+        and isinstance(function, ObjectProxy)
+        and hasattr(function, "__wrapped__")
+    ):
+        setattr(obj, attr, function.__wrapped__)
+
+
+def add_span_arg_tags(span, endpoint_name, args, args_names, args_traced):
+    def truncate_arg_value(value, max_len=1024):
```

Comment:
this is shared with the boto instrumentation. would it make sense to combine the botocore and the boto instrumentation?

Reply:
I think it would be better to have one instrumentation as a dependency of the other one. I'll open an issue to track that.